invisible_cities/evm/histos.py

```
@@ -0,0 +1,194 @@
+
+
+class Histogram:
+    def __init__(self, title, bins, labels, values=None):
+        """
+        This class represents a histogram with is a parameter holder that
+        contains data grouped by bins.
+
+        Attributes:
+        title     = String with the histogram title
+        bins      = List with the histogram binning.
+        data      = Array with the accumulated entries on each bin.
+        out_range = Array with the Accumulated counts out of the bin range.
+                    Values are n-dim arrays of lenght 2 (first element is
+                    underflow, second oveflow).
+        errors    = Array with the assigned uncertanties to each bin.
+        labels    = List with the axis labels.
+
+        Arguments:
+        bins   = List containing the histogram binning.
+        values = Array with initial values, optional.
+                 If not passed, then the initial bin content is set to zero.
+        """
+        self.title     = title
+        self.bins      = bins
+        self.data      = self.init_from_bins()
+        self.errors    = self.init_from_bins()
+        self.out_range = np.zeros(shape=(2, len(self.bins)))
+        self.labels    = labels
+
+        if values is not None:
+            self.fill(np.asarray(values))
+
+    def init_from_bins(self):
+        "Encapsulation for histogram initialization to 0"
+        return np.zeros(shape=tuple(len(x) - 1 for x in self.bins))
+
+    def fill(self, additive, data_weights=[]):
+        """
+        Given datapoints, bins and adds thems to the stored bin content.
+
+        Arguments:
+        additive     = Array or list with data to fill the histogram.
+        data_weights = Array or list with weights of the data.
+        """
+        additive = np.array(additive)
+
+        if len(data_weights) != len(additive.T):
+            data_weights = np.ones (len(additive.T))
+        else:
+            data_weights = np.array(data_weights)
+
```

A few things:
1. The `!=` comparison is to check whether `data_weights` is empty? If so, then just check its size. An invalid length of that sequence should raise an error, I guess.
2. The default value for `data_weights` should be an empty tuple, as the list object is mutable.
3. You can use `a = some_value if condition else other_value` for assignments of this kind. If it is too long, leave it as it is, but put the code in the same line as the `if`/`else`.
4. I find more elegant `additive.shape[1]` than `len(additive.T)`